stdlib/universe/holt_winters/holt_winters.go

```diff
@@ -0,0 +1,289 @@
+
+import (
+	"math"
+
+	"github.com/apache/arrow/go/arrow/array"
+)
+
+// HoltWinters forecasts a series into the future.
+// This is done using the Holt-Winters damped method.
+//    1. The initial values are calculated using a SSE.
+//    2. The series is forecast into the future using the iterative relations.
+type HoltWinters struct {
+	n              int
+	s              int
+	seasonal       bool
+	includeFitData bool
+	// NelderMead optimizer
+	optim *Optimizer
+	// Small difference bound for the optimizer
+	epsilon float64
+
+	vs         *array.Float64
+	newBuilder func() *array.Float64Builder
+}
+
+const (
+	// Arbitrary weight for initializing some intial guesses.
+	// This should be in the  range [0,1]
+	hwWeight = 0.5
+	// Epsilon value for the minimization process
+	hwDefaultEpsilon = 1.0e-4
+	// Define a grid of initial guesses for the parameters: alpha, beta, gamma, and phi.
+	// Keep in mind that this grid is N^4 so we should keep N small
+	// The starting lower guess
+	hwGuessLower = 0.3
+	//  The upper bound on the grid
+	hwGuessUpper = 1.0
+	// The step between guesses
+	hwGuessStep = 0.4
+)
+
+// New creates a new HoltWinters.
+func New(n, s int, withFit bool) *HoltWinters {
+	seasonal := true
+	if s < 2 {
+		seasonal = false
+	}
+	return &HoltWinters{
+		n:              n,
+		s:              s,
+		seasonal:       seasonal,
+		includeFitData: withFit,
+		optim:          NewOptimizer(),
+		epsilon:        hwDefaultEpsilon,
+	}
+}
+
+// Do returns the points generated by the HoltWinters algorithm.
+// Do is given the values on which to compute the algorithm and a function to generate arrow-array builders.
+// This is due to the fact that the algorithm internally creates and releases arrow arrays.
+func (r *HoltWinters) Do(vs *array.Float64, newBuilder func() *array.Float64Builder) *array.Float64 {
+	l := vs.Len() // l is the length of both times and values
+	if l < 2 || r.seasonal && l < r.s || r.n <= 0 {
+		return newBuilder().NewFloat64Array()
+	}
+	r.vs = vs
+	r.newBuilder = newBuilder
+	m := r.s
+
+	// Starting guesses
+	// NOTE: Since these values are guesses
+	// in the cases where we were missing data,
+	// we can just skip the value and call it good.
+	l0 := 0.0
+	if r.seasonal {
+		for i := 0; i < m; i++ {
+			if vs.IsValid(i) {
+				l0 += (1 / float64(m)) * vs.Value(i)
+			}
+		}
+	} else {
+		l0 += hwWeight * vs.Value(0)
+	}
+
+	b0 := 0.0
+	if r.seasonal {
+		for i := 0; i < m && m+i < vs.Len(); i++ {
+			if vs.IsValid(i) && vs.IsValid(m+i) {
+				b0 += 1 / float64(m*m) * (vs.Value(m+i) - vs.Value(i))
+			}
+		}
+	} else {
+		if vs.IsValid(1) {
+			b0 = hwWeight * (vs.Value(1) - vs.Value(0))
+		}
+	}
+
+	// We both want to track memory usage and use slices.
+	// The algorithm changes values of params and this isn't possible with arrow arrays.
+	// So, we build a new arrow array, and we use its underlying slice for manipulation.
+	pb := newBuilder()
+	size := 6
+	if r.seasonal {
+		size += m
+	}
+	pb.Reserve(size)
```

Comment:
it seems like even if `r.seasonal` is true, that `size` is trivially small, so that it's sufficiently constant relative to the rest of the algorithm's memory usage.   Is an arrow array backing worth the overhead involved?

Reply:
Yeah, I was wondering that too. I think you've got the point. `n` should be reasonably small (max 1k values? I think it is a correct assumption to say no-one will need to predict millions of values). I'll turn the implementation back to slices 👍 

I'll update with using `Allocate` for the allocator, just to remove the overhead of arrow, but tracking memory usage. Then we can discuss if that's a good approach or not :)